Archive/Imo/Imo2024Q1.lean

```diff
@@ -0,0 +1,198 @@
+import Mathlib.Tactic.Peel
+
+/-!
+# IMO 2024 Q1
+
+Determine all real numbers $\alpha$ such that, for every positive integer $n$, the integer
+\[
+\lfloor \alpha \rfloor + \lfloor 2\alpha \rfloor + \cdots + \lfloor n\alpha \rfloor
+\]
+is a multiple of~$n$.
+
+We follow Solution 3 from the
+[official solutions](https://www.imo2024.uk/s/IMO-2024-Paper-1-Solutions.pdf).  First reducing
+modulo 2, any answer that is not a multiple of 2 is inductively shown to be contained in a
+decreasing sequence of intervals, with empty intersection.
+-/
+
+
+namespace Imo2024Q1
+
+/-- The condition of the problem. -/
+def Condition (α : ℝ) : Prop := (∀ n : ℕ, 0 < n → (n : ℤ) ∣ ∑ i ∈ Finset.Icc 1 n, ⌊i * α⌋)
+
+lemma condition_two_mul_int (m : ℤ) : Condition (2 * m) := by
+  rintro n -
+  suffices (n : ℤ) ∣ ∑ i ∈ Finset.Icc 1 n, ⌊((i * (2 * m) : ℤ) : ℝ)⌋ from mod_cast this
+  suffices (n : ℤ) ∣ ∑ i ∈ Finset.Icc 0 n, ⌊((i * (2 * m) : ℤ) : ℝ)⌋ by
+    convert this using 1
+    convert (Finset.sum_insert_zero (a := 0) ?_).symm
+    · exact (Nat.Icc_insert_succ_left (Nat.zero_le n)).symm
+    · simp
+  simp_rw [Int.floor_intCast, ← Finset.sum_mul, ← Nat.Ico_succ_right, Nat.succ_eq_add_one,
+           ← Finset.range_eq_Ico, ← mul_assoc]
+  refine dvd_mul_of_dvd_left ?_ _
+  suffices (n : ℤ) ∣ ((n + 1) * (n + 1 - 1) : ℕ) by
+    convert this
+    rw [← Finset.sum_range_id_mul_two]
+    push_cast
+    rfl
+  simp
+
+lemma condition_sub_two_mul_int_iff {α : ℝ} (m : ℤ) : Condition (α - 2 * m) ↔ Condition α := by
+  simp_rw [Condition]
+  peel with n hn
+  refine dvd_iff_dvd_of_dvd_sub ?_
+  simp_rw [← Finset.sum_sub_distrib, mul_sub]
+  norm_cast
+  simp_rw [Int.floor_sub_int, sub_sub_cancel_left]
+  convert condition_two_mul_int (-m) n hn
+  norm_cast
+  rw [Int.floor_intCast]
+  simp
+
+lemma condition_toIcoMod_iff {α : ℝ} :
+    Condition (toIcoMod (by norm_num : (0 : ℝ) < 2) 0 α) ↔ Condition α := by
+  rw [toIcoMod, zsmul_eq_mul, mul_comm, condition_sub_two_mul_int_iff]
+
+namespace Condition
+
+variable {α : ℝ} (hc : Condition α)
+
+lemma mem_Ico_one_of_mem_Ioo (h : α ∈ Set.Ioo 0 2) : α ∈ Set.Ico 1 2 := by
+  rcases h with ⟨h0, h2⟩
+  refine ⟨?_, h2⟩
+  by_contra hn
+  rw [not_le] at hn
+  have hr : 1 < ⌈α⁻¹⌉₊ := by
+    rw [Nat.lt_ceil]
+    exact_mod_cast one_lt_inv h0 hn
+  replace hc := hc ⌈α⁻¹⌉₊ (zero_lt_one.trans hr)
+  refine hr.ne' ?_
+  suffices ⌈α⁻¹⌉₊ = (1 : ℤ) from mod_cast this
+  refine Int.eq_one_of_dvd_one (Int.zero_le_ofNat _) ?_
+  convert hc
+  rw [← Finset.add_sum_Ico_eq_sum_Icc hr.le]
+  convert (add_zero _).symm
+  · rw [Int.floor_eq_iff]
+    refine ⟨?_, ?_⟩
+    · rw [Int.cast_one]
+      calc 1 ≤ α⁻¹ * α := by simp [h0.ne']
+        _ ≤ ⌈α⁻¹⌉₊ * α := by gcongr; exact Nat.le_ceil _
+    · calc ⌈α⁻¹⌉₊ * α < (α⁻¹ + 1) * α := by gcongr; exact Nat.ceil_lt_add_one (inv_nonneg.2 h0.le)
+        _ = 1 + α := by field_simp [h0.ne']
+        _ ≤ (1 : ℕ) + 1 := by gcongr; norm_cast
+  · suffices ∀ x ∈ Finset.Ico 1 ⌈α⁻¹⌉₊, ⌊x * α⌋ = 0 by
```

Comment:
Simpler:
```lean
suffices ∀ x ∈ Finset.Ico 1 ⌈α⁻¹⌉₊, ⌊x * α⌋ = 0 from Finset.sum_eq_zero this
```

Reply:
(or even `refine Finset.sum_eq_zero ?_`)

Reply:
I've made that change.